### PR TITLE
[SPARK-34311][SQL] PostgresDialect can't treat arrays of some types

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR fixes the issue that `PostgresDialect` can't treat arrays of some types.
Though PostgreSQL supports wide range of types (https://www.postgresql.org/docs/13/datatype.html),  the current `PostgresDialect` can't treat arrays of the following types.

* xml
* tsvector
* tsquery
* macaddr
* macaddr8
* txid_snapshot
* pg_snapshot
* point
* line
* lseg
* box
* path
* polygon
* circle
* pg_lsn
* bit varying
* interval

NOTE: PostgreSQL doesn't implement arrays of serial types so this PR doesn't care about them.

### Why are the changes needed?

To provide better support with PostgreSQL.

### Does this PR introduce _any_ user-facing change?

Yes. PostgresDialect can handle arrays of types shown above.

### How was this patch tested?

New test.

### CODE REVIEW

**Comment by sarutak** [comment]
For reviews: Array of series is not implemented in PostgreSQL so I didn't care about `series2`, `series4` and `series8`.

**Review by srowen — APPROVED** [review]
Seems OK if these new types all have a string representation that's meaningful, and it just helps read more data types.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala [hunk 1]
```diff
@@ -74,7 +74,9 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationSuite {
       .executeUpdate()
 
     conn.prepareStatement("CREATE TABLE st_with_array (c0 uuid, c1 inet, c2 cidr," +
-      "c3 json, c4 jsonb, c5 uuid[], c6 inet[], c7 cidr[], c8 json[], c9 jsonb[])")
+      "c3 json, c4 jsonb, c5 uuid[], c6 inet[], c7 cidr[], c8 json[], c9 jsonb[], c10 xml[], " +
```
Comment (maropu) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala:77]: How about adding tests for non-array cases, e.g., `col xml` and `col tsvector`?

**Comment by maropu** [comment]
Does this PR supports all types in PostgreSQL? are there unsupported types?

**Comment by sarutak** [comment]
> Does this PR supports all types in PostgreSQL? are there unsupported types?

I think so except for serial types as I mentioned [above](https://github.com/apache/spark/pull/31419#issuecomment-770883149).

Here is all types in PostgreSQL.
https://www.postgresql.org/docs/13/datatype.html

**Comment by maropu** [comment]
okay, could you copy the statement above into the PR description?

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala (hunk 1 above)
Comment (sarutak) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala:77]: I found there are lots of types not tested so far. So I'll open another PR for non-array cases and focus on array cases in this PR.

**Comment by sarutak** [comment]
I noticed the previous change didn't cover `macaddr8`, `pg_snapshot`, `bit varying` and `interval` so I've updated the code and description.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala (hunk 1 above)
Comment (srowen) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala:77]: Does that mean there are types that aren't supported, but for which we support arrays? did I misunderstand that?

Comment (sarutak) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala:77]: I mean that some supported non-array types like `numeric`, `varchar`, `date` and etc are not tested.
So, we need to add tests for those types in addition to the types supported by this PR.

**Comment by sarutak** [comment]
retest this please.

**Comment by sarutak** [comment]
@srowen @maropu Do you have any comment on the latest change?

**Review by srowen — COMMENTED** [review]
Just so I'm clear, you're saying things like `tsvector` are already supported, and now tested, and now `tsvector[]` is also supported? If not then I'm still wondering why the non-array types can't be supported the same way.

**Comment by sarutak** [comment]
> Just so I'm clear, you're saying things like `tsvector` are already supported, and now tested, and now `tsvector[]` is also supported?

Yes. `tsvector` is already supported but it's not tested so I opened #31456.
On the other hand, `tsvector[]` and other array types added in this PR are not supported before.

**Comment by sarutak** [comment]
@srowen 
Ah, you mean how about mapping to `StringType` by default for array elements?

**Comment by srowen** [comment]
No i just wondered if I should see tests for the non-array types here. It is OK if it's in another PR. This seems fine if you're ready to merge.

**Review by maropu — APPROVED** [review]
I've checked the test can pass on my local env.

**Comment by maropu** [comment]
Thanks! Merged to master.

**Pull request closed without merging**